numpy/core/numeric.py

```diff
@@ -2379,6 +2379,11 @@ def within_tol(x, y, atol, rtol):
 def _array_equal_dispatcher(a1, a2, equal_nan=None):
     return (a1, a2)
 
+def _dtype_cannot_hold_nan(dtype):
+    # can't extract set in module variable as it create a circular import.
+    # AttributeError: partially initialized module 'numpy' has no attribute 'int8' (most likely due to a circular import)
+    return dtype.type in {np.int8, np.int16, np.int32, np.int64}
```

Comment:
```suggestion
_no_nan_types = None
def _dtype_cannot_hold_nan(dtype):
    global _no_nan_types
    if _no_nan_types is None:
        _no_nan_types = {np.bool_, np.int8, np.int16, np.int32, np.int64}
    return dtype.type in _no_nan_types
```
This avoids the circular import.